### PR TITLE
Re-introduce global theme apis

### PR DESCRIPTION
**Pull request by Brailor**
This PR adds a new package called: `theme-registry` which contains a slimmed version of the `ThemeRegistry` from v7.
This solution enables hybrid theme registration, users can apply themes 2 different ways:
- wrap their React App roots with our `InstUISettingsprovider`
- or import the theme they want to use directly and call `theme.use()` on it:
```ts
import { canvas, canvasHighContrast }  from '@instructure/ui-themes'

if(someCondition) {
    canvasHighContrast.use()
} else {
    canvas.use()
}
...
```

It is also possible to override themes that are being `use()`-d:
```ts
canvas.use({
    overrides: {
        colors: {
             brand: 'red'
        }
    }
})
```

It is possible to use both solutions together, so lets say an application sets a global `canvas` theme, with the `canvas.use({ overrides:... })`, then later down the app tree they wrap a part of the sub-tree with an `InstUISettingsprovider theme={canvasHighContrast}`, then the context based theming will take precedence.

### CODE REVIEW

File: docs/guides/using-theme-overrides.md [hunk 1]
```diff
@@ -36,11 +39,17 @@ embed: true
...
 ### How theming works in InstUI
 
-The theming system in InstUI has two levels:
+The theming system in InstUI has three levels:
 
 **The global theme**
 
-On the broader level, there is the main theme object that contains the color, spacing, typography etc. variables available in the theme (e.g.: [canvas theme](/#canvas)). The global theme can be set via the [InstUISettingsProvider](/#InstUISettingsProvider) component.
+Global themes are useful when you have multiple React Application trees and you want to set up themes and overrides on a more global level than application level theming.
+This basically means you don't necesseraly have to warp each application tree with an [InstUISettingsProvider](/#InstUISettingsProvider) to use themes.
```
Comment (matyasf) [inline docs/guides/using-theme-overrides.md:47]: typo: ... **necessarily** have to **wrap**...

File: docs/guides/using-theme-overrides.md [hunk 2]
```diff
@@ -421,3 +461,139 @@ example: true
...
+})
+```
+
+This will register the `canvas` theme with its overrides as the active theme in the [ThemeRegistry](/#ThemeRegistry) and then every InstUI component - that is not wrapped in an `InstUISettingsProvider` - will use this theme no matter which React tree they are rendered in.
+
+#### Use global theme together with InstUISettingsProvider
+
+You can use the global theming and the application level theming together, however there are some caveats you should be aware of:
+
+- if you globally `.use()` a theme and would like to override that theme in a sub-tree of your application, then make sure that sub-tree does not have a theme provided to any `InstUISettingsProvider` above it in the tree:
```
Comment (matyasf) [inline docs/guides/using-theme-overrides.md:540]: hmm this part is a bit hard to read.. I'd rephrase it like this (is my interpretation correct?):
"if you globally `.use()` a theme and would like to override that theme in a sub-tree of your application then just wrap that part in `InstUISettingsProvider`. Here just specify the overrides, not the full theme:"

File: package.json [hunk 3]
```diff
@@ -15,8 +15,8 @@
   ],
   "scripts": {
     "prestart": "yarn bootstrap",
-    "start": "lerna run start --stream --scope docs-app",
-    "start:watch": "lerna run start:watch --stream --scope docs-app",
+    "start": "yarn workspace docs-app start",
+    "start:watch": "yarn workspace docs-app start:watch ",
```
Comment (matyasf) [inline package.json:19]: less lerna is always good :)

File: packages/theme-registry/src/ThemeRegistry.ts [hunk 4]
```diff
@@ -0,0 +1,267 @@
...
+  const wrappedTheme = {
+    key,
+    description,
+    ...rest,
+    use(arg?: { overrides: DeepPartial<BaseThemeVariables> }): void {
+      activateTheme(key, arg?.overrides || {})
+    }
+  }
+
+  return new Proxy(wrappedTheme, {
```
Comment (matyasf) [inline packages/theme-registry/src/ThemeRegistry.ts:214]: nice, first time I see `Proxy` used in code :)

**Review by matyasf — APPROVED** [review]
looks good (just some small text adjustments), hope Canvas likes it

File: packages/theme-registry/README.md [hunk 5]
```diff
@@ -0,0 +1,25 @@
...
+category: packages
+---
+
+## theme-registry
+
+[![npm][npm]][npm-url]
+[![MIT License][license-badge]][license]
+[![Code of Conduct][coc-badge]][coc]
+
+A pacakge for managing global themes.
```
Comment (ke1k0) [inline packages/theme-registry/README.md:11]:
```suggestion
A package for managing global themes.
```

File: packages/theme-registry/src/ThemeRegistry.ts [hunk 6]
```diff
@@ -0,0 +1,267 @@
...
+ * and recieving the currently used theme.
+ * @module ThemeRegistry
+ */
+
+import { error } from '@instructure/console'
+import {
+  BaseTheme,
+  BaseThemeVariables,
+  DeepPartial
+} from '@instructure/shared-types'
```
Comment (ke1k0) [inline packages/theme-registry/src/ThemeRegistry.ts:39]:
```suggestion
import type {
  BaseTheme,
  BaseThemeVariables,
  DeepPartial
} from '@instructure/shared-types'
```

File: packages/theme-registry/src/ThemeRegistry.ts [hunk 7]
```diff
@@ -0,0 +1,267 @@
...
+ * @return {Registry}  the theme registry
+ * @module getRegistry
+ */
+function getRegistry(): Registry<RegisteredTheme> {
+  return globalThis[GLOBAL_THEME_REGISTRY]
+}
+
+/**
+ * Set the global theme registry.
+ * @param {Registry} registry - the registry to set/replacet the current registry with.
```
Comment (ke1k0) [inline packages/theme-registry/src/ThemeRegistry.ts:131]:
```suggestion
 * @param {Registry} registry - the registry to set/replace the current registry with.
```

File: packages/theme-registry/src/ThemeRegistry.ts [hunk 8]
```diff
@@ -0,0 +1,267 @@
...
+
+  return themeWithOverrides
+}
+/**
+ * Wraps a theme and provides a method to activate the theme.
+ * @returns the wrapped theme object
+ * @module makeTheme
+ * @private
+ */
+function makeTheme<T extends BaseTheme>(theme: T): RegisteredTheme<T> {
```
Comment (ke1k0) [inline packages/theme-registry/src/ThemeRegistry.ts:203]: Do you think we need some checks here that check if the theme object is a valid instui theme format?

**Review by ke1k0 — COMMENTED** [review]
I tested it as much as I could, seems to work fine. I found a few typos and commented a few questions too :) Nice work!

File: docs/guides/using-theme-overrides.md (hunk 1 above)
Comment (Brailor) [reply docs/guides/using-theme-overrides.md]: fixed

File: docs/guides/using-theme-overrides.md [hunk 9]
```diff
@@ -15,6 +15,9 @@ order: 4
 - [Component themeOverride prop](/#using-theme-overrides/#using-theme-overrides-component-themeoverride-prop)
   - [Override object](/#using-theme-overrides/#using-theme-overrides-component-themeoverride-prop-override-object)
   - [Override function](/#using-theme-overrides/#using-theme-overrides-component-themeoverride-prop-override-function)
+- [Examples](/#using-theme-overrides/#using-theme-overrides-examples)
+  - [Use global theme for multiple React trees](/#using-theme-overrides/#using-theme-overrides-examples-use-global-theme-for-multiple-react-trees)
+  - [Use global theme together with InstUISettingsProvider](/#using-theme-overrides/#using-theme-overrides-examples-use-global-theme-together-with-instuisettingsprovider)
```
[the remark this answers is not in the document]
Comment (Brailor) [reply docs/guides/using-theme-overrides.md]: I've deleted the examples TOC, thanks

File: docs/guides/using-theme-overrides.md [hunk 10]
```diff
@@ -421,3 +461,139 @@ example: true
   </div>
 </div>
 ```
+
+### Examples
```
[the remark this answers is not in the document]
Comment (Brailor) [reply docs/guides/using-theme-overrides.md]: Changed

File: docs/guides/using-theme-overrides.md [hunk 11]
```diff
@@ -421,3 +461,139 @@ example: true
...
+
+#### Use global theme for multiple React trees
+
+If you have multiple `ReactDOM.render` calls and you would like to use the same theme with the same overrides for them, then you have 2 options:
+
+- wrap each `ReactDOM.render` application with an `InstUISettingsProvider` with the correct theme:
```
[the remark this answers is not in the document]
Comment (Brailor) [reply docs/guides/using-theme-overrides.md]: Changed to sub titles

File: docs/guides/using-theme-overrides.md [hunk 12]
```diff
@@ -36,11 +39,17 @@ embed: true
...
 
-The theming system in InstUI has two levels:
+The theming system in InstUI has three levels:
 
 **The global theme**
 
-On the broader level, there is the main theme object that contains the color, spacing, typography etc. variables available in the theme (e.g.: [canvas theme](/#canvas)). The global theme can be set via the [InstUISettingsProvider](/#InstUISettingsProvider) component.
+Global themes are useful when you have multiple React Application trees and you want to set up themes and overrides on a more global level than application level theming.
+This basically means you don't necesseraly have to warp each application tree with an [InstUISettingsProvider](/#InstUISettingsProvider) to use themes.
+InstUI leverages the [ThemeRegistry](/#ThemeRegistry) package to achive global theming.
```
[the remark this answers is not in the document]
Comment (Brailor) [reply docs/guides/using-theme-overrides.md]: Fixed

File: docs/guides/using-theme-overrides.md [hunk 13]
```diff
@@ -49,7 +58,7 @@ Every themeable component has its own "theme map". This map defines the componen
 See the [emotion](/#emotion) and [InstUISettingsProvider](/#InstUISettingsProvider) docs pages for more info and examples.
 
 ```jsx
-// app/component root sets the global theme
+// app/component root sets the app theme
```
[the remark this answers is not in the document]
Comment (Brailor) [reply docs/guides/using-theme-overrides.md:69]: I added an example, please check it, not sure this is what you had in mind

File: docs/guides/using-theme-overrides.md [hunk 14]
```diff
@@ -421,3 +461,139 @@ example: true
...
+  ReactDOM.render(
+    <InstUISettingsProvider theme={themeWithOverrides}>
+      <App2>
+    </InstUISettingsProvider>
+  )
+  ```
+
+  This approach quickly becomes too cumbersome if you have more than a couple of `ReactDOM.render` calls.
+
+- instead you could achive the same thing with global themes:
```
[the remark this answers is not in the document]
Comment (Brailor) [reply docs/guides/using-theme-overrides.md]: I ended up rephrasing that line

File: docs/guides/using-theme-overrides.md [hunk 15]
```diff
@@ -421,3 +461,139 @@ example: true
...
+        }
+      }}>
+      //Any InstUI component that uses the `theme.colors.brand` will have it's color set to "magenta" in this subtree
+      <SubTree />
+    </InstUISettingsProvider>
+    )
+  }
+  ```
+
+- this example won't override the global theme, instead it will the override the application level theme provided to `InstUISettingsProvider`
```
[the remark this answers is not in the document]
Comment (Brailor) [reply docs/guides/using-theme-overrides.md]: Fixed

File: docs/guides/using-theme-overrides.md [hunk 16]
```diff
@@ -421,3 +461,139 @@ example: true
...
+  ```jsx
+  //your application root code:
+  import { instructure } from "@instructure/ui-themes"
+
+  render() {
+    return (
+      <InstUISettingsProvider theme={instructure}>
+        <SubTree1 />
+        //This theme override will only change the theme that is directly above it in the context,
+        //the 'instructure' theme not the global 'canvas' theme!
```
[the remark this answers is not in the document]
Comment (Brailor) [reply docs/guides/using-theme-overrides.md]: I deleted the comment

File: packages/canvas-high-contrast-theme/README.md [hunk 17]
```diff
@@ -22,30 +22,50 @@ yarn add @instructure/canvas-high-contrast-theme
...
-import { theme } from '@instructure/canvas-high-contrast-theme'
-
-ReactDOM.render(
-  <InstUISettingsProvider theme={theme}>
-    <App />
-  </InstUISettingsProvider>,
-  element
-)
-```
+- global theming:
```
[the remark this answers is not in the document]
Comment (Brailor) [reply packages/canvas-high-contrast-theme/README.md:25]: Done

File: packages/theme-registry/CHANGELOG.md [hunk 18]
```diff
@@ -0,0 +1,314 @@
+# Change Log
+
+All notable changes to this project will be documented in this file.
+See [Conventional Commits](https://conventionalcommits.org) for commit guidelines.
+
+# [8.20.0](https://github.com/instructure/instructure-ui/compare/v8.19.0...v8.20.0) (2022-03-22)
```
[the remark this answers is not in the document]
Comment (Brailor) [reply packages/theme-registry/CHANGELOG.md]: I didn't know this one, thanks, deleted all its contents

File: packages/theme-registry/README.md (hunk 5 above)
Comment (Brailor) [reply packages/theme-registry/README.md]: Fixed

File: packages/theme-registry/src/ThemeRegistry.ts (hunk 6 above)
Comment (Brailor) [reply packages/theme-registry/src/ThemeRegistry.ts]: Fixed

File: packages/theme-registry/src/ThemeRegistry.ts (hunk 7 above)
Comment (Brailor) [reply packages/theme-registry/src/ThemeRegistry.ts]: Fixed

File: packages/theme-registry/src/ThemeRegistry.ts (hunk 8 above)
Comment (Brailor) [reply packages/theme-registry/src/ThemeRegistry.ts:203]: I moved the `isBaseTheme` from `getTheme.ts` and used that function to validate the theme.

**Review by ke1k0 — APPROVED** [review]
I left a few "nitpicking" comments, so nothing crucial, but I'm giving a +1 :) Nice work!

File: packages/theme-registry/src/ThemeRegistry.ts (hunk 8 above)
Comment (Brailor) [reply packages/theme-registry/src/ThemeRegistry.ts:203]: Yeah, this one is not easy. I mean you could register themes that don't have all the expected properties, but then our components could behave weirdly because they expect all the properties to be there.

I think I like the check more, and if for some reason we have to delete this check later, then it won't be a breaking change, however if we don't add the check now and have to add it later, then that would be a breaking change.

Comment (ke1k0) [reply packages/theme-registry/src/ThemeRegistry.ts:203]: Very too! I prefer having this check too, tbh :)

File: docs/guides/using-theme-overrides.md [hunk 19]
```diff
@@ -421,3 +473,139 @@ example: true
...
+})
+```
+
+This will register the `canvas` theme with its overrides as the active theme in the [ThemeRegistry](/#ThemeRegistry) and then every InstUI component - that is not wrapped in an `InstUISettingsProvider` - will use this theme no matter which React tree they are rendered in.
+
+#### Use global theme together with InstUISettingsProvider
+
+You can use the global theming and the application level theming together, however there are some caveats you should be aware of:
+
+##### If you globally .use() a theme and would like to override that theme in a sub-tree of your application then just wrap that part in an `InstUISettingsProvider`
```
[the remark this answers is not in the document]
Comment (Brailor) [reply docs/guides/using-theme-overrides.md]: > I left a few "nitpicking" comments, so nothing crucial, but I'm giving a +1 :) Nice work!
Thanks, I made all of the suggested changes.
